### ISSUE TITLE
QgsMapCanvas.mapToolSet() signal wrong parameter types

### ISSUE THREAD

**Issue by folini96** [qgis/QGIS#50057]
### What is the bug or the crash?

The signal QgsMapCanvas.mapToolSet() parameters newTool and oldTool should always have type QgsMapTool. This work for most of the Map Tools, but has some problems with the Selection Map Tools (Select Feature(s) and Select Features by Polygon/Freehand/Radius) which are passed with type QObject. 

### Steps to reproduce the issue

1. Connect the signal to a function
2. In the function print the type name of both parameters
3. Check in the QGIS GUI the different types by selecting the different map tools

### Versions

3.22.10

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by elpaso** [comment]
The reason for this is that most map tools are not in 'gui' library but in 'app' (which is a bug IMHO but not trivial to fix).
I don't see a solution for now. We should probably add this as an API target for QGIS 4.

**Comment by 3nids** [comment]
how a tool can be set if it's not a QgsMapTool? I don't see the issue here.

**Comment by elpaso** [comment]
@3nids to my understanding, the issue is here: https://github.com/qgis/QGIS/blob/master/src/gui/qgsmaptool.h#L74

But that's in `gui` while the missing maptools are in `app`.

**Comment by 3nids** [comment]
I believe the issue was that the sip value is set to null pointer.
see #50303

**Comment by elpaso** [comment]
> I believe the issue was that the sip value is set to null pointer. see #50303

Yes, but that won't fix the fact that for QgsMapToolSelect and other `app` tools the returned type will be the base type QgsMapTool and not the subclass, correct?

**Comment by 3nids** [comment]
yes, but how is this an issue?
looking at the issue here, I believe the point is to get a pointer to the base map tool, not the implementation one, no?

**Comment by elpaso** [comment]
> yes, but how is this an issue? looking at the issue here, I believe the point is to get a pointer to the base map tool, not the implementation one, no?

I honestly don't know, my understanding is that the expected results was to get the implementation one but I might be wrong.

**Comment by folini96** [comment]
The problem is that I was saving the oldTool passed from the signal QgsMapCanvas.mapToolSet() to set it back as the active map tool after I completed an action. This caused a Type Error with the mentioned map tools as they were passed in the oldTool parameter as an instance of QObject (instead of base QgsMapTool which is the expected behavior). I avoided the problem by saving the map tool using mapCanvas.mapTool() before changing it, and in this case all the map tools worked fine. I just wanted to notify the issue since it seems that the signal is not working correctly with some of the map tools implementations.
